Add fw_grad field and accessors

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43718 make jvp use forward mode by default
* #43717 Add gradcheck for forward grad
* #43716 Add many forward grad
* #43715 Basic auto formula for forward grad
* #43714 make sure SavedVariable propagate fw_grad properly
* #43713 Add forward grad mode in python and c++
* **#43712 Add fw_grad field and accessors**

Differential Revision: [D23454142](https://our.internmc.facebook.com/intern/diff/D23454142)

albanD committed:
Add fw_grad field and accessors

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -682,6 +723,7 @@ static struct PyGetSetDef THPVariable_properties[] = {
   {"data", (getter)THPVariable_get_data, (setter)THPVariable_set_data, nullptr, nullptr},
   {"_grad", (getter)THPVariable_get_grad, (setter)THPVariable_set_grad, nullptr, nullptr}, // Allows the python class to override .grad
   {"grad", (getter)THPVariable_get_grad, (setter)THPVariable_set_grad, nullptr, nullptr},
+  {"_fw_grad", (getter)THPVariable_get_fw_grad, (setter)THPVariable_set_fw_grad, nullptr, nullptr},
```
The final API for this will be a `fw_grad` field and a `set_fw_grad(value)` function.
The first one can be used to access, set and reset the value for the forward grad.
Unfortunately, because of the constraints on `[Forward Grad Layout]`, the Tensor that gets set might not be the one that is given.
So setting using the attribute will only be allowed when the given Tensor can actually be used. If this should be an error or a warning will be decided later.
Otherwise, you can use the `set_fw_grad()` function that will return the Tensor that was actually set.

albanD committed:
Update on "Add fw_grad field and accessors"




[ghstack-poisoned]

zou3519 commented on torch/_tensor_str.py:
```diff
@@ -354,6 +354,9 @@ def _str_intern(self):
     if self.layout != torch.strided:
         suffixes.append('layout=' + str(self.layout))
 
+    if self._fw_grad is not None:
+        suffixes.append('_fw_grad={}'.format(self._fw_grad))
```
Add a test for this? (to make sure it prints OK)

albanD committed:
Update on "Add fw_grad field and accessors"




[ghstack-poisoned]

albanD committed:
Update on "Add fw_grad field and accessors"




[ghstack-poisoned]

albanD committed:
Update on "Add fw_grad field and accessors"




[ghstack-poisoned]

ezyang commented on aten/src/ATen/templates/TensorBody.h:
```diff
@@ -552,6 +552,26 @@ class CAFFE2_API Tensor {
+  /// This function can be used to set the value of the forward grad.
+  /// Note that the given value might not be used directly if this Tensor already
+  /// has a forward grad or if it is a view of another Tensor.
+  void set_fw_grad(Tensor& new_grad) {
```
Should we really be encouraging users to mutably set the forward gradient here? It seems better to provide a functional interface by default for setting these values.

albanD replied:
Do you mean the argument should be a `const Tensor&`? Yes definitely a oversight while writing the function!

ezyang replied:
Not just that; if I want to construct a dual number, the default way of doing so should be via some *constructor* that (non-mutably) constructs the dual number, rather than by taking an existing tensor and then mutating it so that it has a `fw_grad`.

I realize you have discussed this in the design in this sentence:

> In the first iteration of this feature with no “user friendly” API, when you want to compute J v, you need to set t.fw_grad = v, then perform your computations. You can then read on the output Tensor out.fw_grad that will contain the result of this computation.

But I think this is a step too far, especially given that modern idiomatic PyTorch usage of `requires_grad` involves setting it in the constructor, rather than as a mutable operation after the fact.

albanD committed:
Update on "Add fw_grad field and accessors"




[ghstack-poisoned]

albanD committed:
Update on "Add fw_grad field and accessors"




[ghstack-poisoned]

albanD committed:
Update on "Add fw_grad field and accessors"




[ghstack-poisoned]

albanD committed:
Update on "Add fw_grad field and accessors"




[ghstack-poisoned]

albanD committed:
Update on "Add fw_grad field and accessors"




[ghstack-poisoned]

albanD committed:
Update on "Add fw_grad field and accessors"




[ghstack-poisoned]

albanD committed:
Update on "Add fw_grad field and accessors"




[ghstack-poisoned]

albanD committed:
Update on "Add fw_grad field and accessors"


Differential Revision: [D23454142](https://our.internmc.facebook.com/intern/diff/D23454142)

[ghstack-poisoned]
